src/platform/input/game-pads.js

```diff
@@ -68,62 +166,763 @@ const MAPS = {
             'PAD_R_STICK_X',
             'PAD_R_STICK_Y'
-        ]
+        ],
+
+        mapping: 'standard'
+    },
+
+    DEFAULT_XR: {
+        buttons: [
+            // Back buttons
+            'XRPAD_TRIGGER',
+            'XRPAD_SQUEEZE',
+
+            // Axes buttons
+            'XRPAD_TOUCHPAD_BUTTON',
+            'XRPAD_STICK_BUTTON',
+
+            // Face buttons
+            'XRPAD_A',
+            'XRPAD_B'
+        ],
+
+        axes: [
+            // Analog Sticks
+            'XRPAD_TOUCHPAD_X',
+            'XRPAD_TOUCHPAD_Y',
+            'XRPAD_STICK_X',
+            'XRPAD_STICK_Y'
+        ],
+
+        mapping: 'xr-standard'
     }
 };
 
 const PRODUCT_CODES = {
     'Product: 0268': 'PS3'
 };
 
+const custom_maps = {};
+
+/**
+ * Retrieve the order for buttons and axes for given HTML5 Gamepad.
+ *
+ * @param {Gamepad} pad - The HTML5 Gamepad object.
+ * @returns {object} Object defining the order of buttons and axes for given HTML5 Gamepad.
+ */
+function getMap(pad) {
+    const custom = custom_maps[pad.id];
+    if (custom) {
+        return custom;
+    }
+
+    for (const code in PRODUCT_CODES) {
+        if (pad.id.indexOf(code) !== -1) {
+            const product = PRODUCT_CODES[code];
+
+            if (!pad.mapping) {
+                const raw = MAPS['RAW_' + product];
+
+                if (raw) {
+                    return raw;
+                }
+            }
+
+            return MAPS[product];
+        }
+    }
+
+    if (pad.mapping === 'xr-standard') {
+        return MAPS.DEFAULT_XR;
+    }
+
+    const defaultmap = MAPS.DEFAULT;
+    const map = pad.buttons.length < defaultmap.buttons.length ? MAPS.DEFAULT_DUAL : defaultmap;
+    map.mapping = pad.mapping;
+    return map;
+}
+
+let deadZone = 0.25;
+
+/**
+ * @param {number} ms - Number of milliseconds to sleep for.
+ * @returns {Promise<void>}
+ */
+function sleep(ms) {
+    return new Promise((resolve) => {
+        setTimeout(resolve, ms);
+    });
+}
+
+/**
+ * A GamePadButton stores information about a button from the Gamepad API.
+ */
+class GamePadButton {
+    /**
+     * The value for the button between 0 and 1, with 0 representing a button that is not pressed, and 1 representing a button that is fully pressed.
+     *
+     * @type {number}
+     */
+    value;
+
+    /**
+     * Whether the button is currently down.
+     *
+     * @type {boolean}
+     */
+    pressed;
+
+    /**
+     * Whether the button is currently touched.
+     *
+     * @type {boolean}
+     */
+
+    touched;
+
+    /**
+     * Whether this button was pressed on last frame.
+     *
+     * @type {boolean}
+     * @ignore
+     */
+    _previouslyPressed;
+
+    /**
+     * Whether this button was touched on last frame.
+     *
+     * @type {boolean}
+     * @ignore
+     */
+    _previouslyTouched;
+
+    /**
+     * The original Gamepad API gamepad button.
+     *
+     * @type {GamepadButton}
+     * @ignore
+     */
+    _button;
```

Comment:
We can get rid of `_button` now after you fixed the Firefox polling issue? Makes the shape of the class a bit smaller + less lines

Reply:
It's done!